nilearn/glm/tests/test_utils.py

```diff
@@ -31,12 +37,101 @@ def test_full_rank():
 
 
 def test_z_score():
-    p = np.random.RandomState(42).uniform(size=10)
-    assert_array_almost_equal(norm.sf(z_score(p)), p)
-    # check the numerical precision
-    for p in [1.e-250, 1 - 1.e-16]:
-        assert_array_almost_equal(z_score(p), norm.isf(p))
-    assert_array_almost_equal(z_score(np.float32(1.e-100)), norm.isf(1.e-300))
+    # ################# Check z-scores computed from t-values #################
+    # Randomly draw samples from the standard Student’s t distribution
+    tval = np.random.standard_t(1e10, size=10)
+    # Estimate the p-values using the Survival Function (SF)
+    pval = sps.t.sf(tval, 1e10)
+    # Estimate the p-values using the Cumulative Distribution Function (CDF)
+    cdfval = sps.t.cdf(tval, 1e10)
+    # Set a minimum threshold for p-values to avoid infinite z-scores
+    pval = np.array(np.minimum(np.maximum(pval, 1.e-300), 1. - 1.e-16))
+    cdfval = np.array(np.minimum(np.maximum(cdfval, 1.e-300), 1. - 1.e-16))
+    # Compute z-score from the p-value estimated with the SF
+    zval_sf = norm.isf(pval)
+    # Compute z-score from the p-value estimated with the CDF
+    zval_cdf = norm.ppf(cdfval)
+    # Create the final array of z-scores, ...
+    zval = np.zeros(pval.size)
+    # ... in which z-scores < 0 estimated w/ SF are replaced by z-scores < 0
+    # estimated w/ CDF
+    zval[np.atleast_1d(zval_sf < 0)] = zval_cdf[zval_sf < 0]
+    # ... and z-scores >=0 estimated from SF are kept.
+    zval[np.atleast_1d(zval_sf >= 0)] = zval_sf[zval_sf >= 0]
+    # Test 'z_score' function in 'nilearn/glm/contrasts.py'
+    assert_array_almost_equal(z_score(pval, one_minus_pvalue=cdfval), zval)
+    # ################# Check z-scores computed from F-values #################
+    # Randomly draw samples from the F distribution
+    fval = np.random.f(1e10, 42, size=10)
+    # Estimate the p-values using the Survival Function (SF)
+    p_val = sps.f.sf(fval, 42, 1e10)
+    # Estimate the p-values using the Cumulative Distribution Function (CDF)
+    cdf_val = sps.f.cdf(fval, 42, 1e10)
+    # Set a minimum threshold for p-values to avoid infinite z-scores
+    p_val = np.array(np.minimum(np.maximum(p_val, 1.e-300), 1. - 1.e-16))
+    cdf_val = np.array(np.minimum(np.maximum(cdf_val, 1.e-300), 1. - 1.e-16))
+    # Compute z-score from the p-value estimated with the SF
+    z_val_sf = norm.isf(p_val)
+    # Compute z-score from the p-value estimated with the CDF
+    z_val_cdf = norm.ppf(cdf_val)
+    # Create the final array of z-scores, ...
+    z_val = np.zeros(p_val.size)
+    # ... in which z-scores < 0 estimated w/ SF are replaced by z-scores < 0
+    # estimated w/ CDF
+    z_val[np.atleast_1d(z_val_sf < 0)] = z_val_cdf[z_val_sf < 0]
+    # ... and z-scores >=0 estimated from SF are kept.
+    z_val[np.atleast_1d(z_val_sf >= 0)] = z_val_sf[z_val_sf >= 0]
+    # Test 'z_score' function in 'nilearn/glm/contrasts.py'
+    assert_array_almost_equal(z_score(p_val, one_minus_pvalue=cdf_val), z_val)
```

Comment:
this doesn't seem to really test if the results are correct, as it compares them with the output of code that is almost identical to that of the function itself?

Reply:
What would be an alternative solution?

Reply:
the comparison with opposite contrast (below) seems to be a more sensible check -- and it is how the issue was initially reported, right? so maybe just remove this test?

Reply:
But this test function tests whether the `z_score` function from `nilearn/_utils/glm.py` is working. It was already here before. I had to change it because I changed the `z_score` function and then this test function started producing an error. I had thus to adapt it to the present context.